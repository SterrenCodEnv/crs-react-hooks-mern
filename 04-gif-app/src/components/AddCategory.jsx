import { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState('');

  const onInputChange = ({ target: { value } }) => {
    setInputValue(value);
  };

  const onSubmit = (e) => {
    e.preventDefault();
    if (inputValue.length >= 1) {
      onNewCategory(inputValue.trim().toLocaleLowerCase());
      setInputValue('');
    }
  };

  return (
    <form onSubmit={onSubmit} aria-label="form">
      <input
        onChange={onInputChange}
        value={inputValue}
        type="text"
        placeholder="Buscar gif..."
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
