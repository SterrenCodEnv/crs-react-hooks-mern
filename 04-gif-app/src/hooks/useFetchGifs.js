import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/GetGifs';

export const useFetchGifs = (category) => {
  const [images, setImages] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    setTimeout(() => {
      getGifs(category).then(setImages);
      setIsLoading(false);
    }, 500);
  }, []);

  return {
    images,
    isLoading,
  };
};
