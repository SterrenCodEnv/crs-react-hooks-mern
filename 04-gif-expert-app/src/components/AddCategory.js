import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const hanleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const hadleSubmit = (e) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) setCategories((c) => [inputValue, ...c]);
    setInputValue("");
  };

  return (
    <>
      <form onSubmit={hadleSubmit}>
        <input type="text" value={inputValue} onChange={hanleInputChange} />
      </form>
    </>
  );
};

AddCategory.prototype = {
  setCategories: PropTypes.func.isRequired,
};
