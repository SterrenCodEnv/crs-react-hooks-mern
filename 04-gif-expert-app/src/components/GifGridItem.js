import React from 'react';
import PropTypes from 'prop-types';
const heroImages = require.context('../../assets/images', true);

export const GifGridItem = ({ title, url }) => {
  return (
    <div className='card animate__animated animate__fadeIn'>
      <img src={heroImages(`./${url}.jpg`)} alt={title}></img>
      <p>{title}</p>
    </div>
  );
};

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
};
