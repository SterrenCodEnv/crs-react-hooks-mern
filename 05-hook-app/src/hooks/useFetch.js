import { useEffect, useState, useRef } from "react";

export const useFetch = (url) => {
  const isMount = useRef(true);

  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    return () => {
      isMount.current = false;
    };
  }, []);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url).then((resp) => {
      resp
        .json()
        .then((data) => {
          if (isMount.current) {
            setState({
              data,
              loading: false,
              error: null,
            });
          }
        })
        .catch(() => {
          setState({
            data: null,
            loading: false,
            error: "Data could not be loaded",
          });
        });
    });
  }, [url]);

  return state;
};
