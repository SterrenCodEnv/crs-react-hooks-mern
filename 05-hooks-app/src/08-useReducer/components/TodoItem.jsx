export const TodoItem = ({ todo, onDeleteTodo, onToggleTodo }) => {
  return (
    <>
      <li className="list-group-item d-flex justify-content-between">
        <span
          style={{ cursor: 'pointer' }}
          className={`align-self-center unselect text-capitalize ${
            todo.completed
              ? 'text-decoration-line-through text-muted fst-italic'
              : ''
          }`}
          onClick={() => onToggleTodo(todo.id)}
        >
          {todo.title}
        </span>
        <button
          onClick={() => onDeleteTodo(todo.id)}
          className="btn btn-sm btn-danger btn-circle"
        >
          <i className="fa-solid fa-trash"></i>
        </button>
      </li>
    </>
  );
};
