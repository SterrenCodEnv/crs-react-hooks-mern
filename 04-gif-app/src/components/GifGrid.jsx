import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifItem } from './GifItem';

export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);

  return (
    <>
      <div key={category}>
        <h2>{`${
          category.charAt(0).toUpperCase() + category.slice(1).toLowerCase()
        }`}</h2>

        {isLoading && <h2>Loading...</h2>}

        <div className="card-grid">
          {images.map((image) => (
            <GifItem key={image.id} {...image} />
          ))}
        </div>
      </div>
    </>
  );
};
