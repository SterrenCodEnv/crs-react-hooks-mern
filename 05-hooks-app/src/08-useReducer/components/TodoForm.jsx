import { useForm } from '../../hooks/useForm';

export const TodoForm = ({ onNewTodo }) => {
  const { title, onInputChange, onResetForm } = useForm({
    title: '',
  });

  const onSubmit = (e) => {
    e.preventDefault();
    if (title.trim() === '') return;
    const newTodo = {
      id: new Date().getTime(),
      title,
      completed: false,
    };
    onNewTodo(newTodo);
    onResetForm();
  };

  return (
    <>
      <form onSubmit={onSubmit}>
        <input
          name="title"
          type="text"
          placeholder="What's there to do?"
          className="form-control"
          value={title}
          onChange={onInputChange}
        />
        <button type="submit" className="btn btn-outline-primary mt-3">
          Add
        </button>
      </form>
    </>
  );
};
