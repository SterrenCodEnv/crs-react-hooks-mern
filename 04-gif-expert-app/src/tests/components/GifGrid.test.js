import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFecthGifs";
jest.mock("../../hooks/useFecthGifs");

describe("Test in GifGrid", () => {
  const category = "The Office";

  test("Show GifGrid component correctly", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Show items when loading images in useFetchGifs", () => {
    const gifs = [
      {
        id: "abc",
        ulr: "https://google.com",
        title: "Gif Title",
      },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
