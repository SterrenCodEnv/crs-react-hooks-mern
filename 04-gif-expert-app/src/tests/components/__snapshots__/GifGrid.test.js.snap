// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in GifGrid Show GifGrid component correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    The Office
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Test in GifGrid Show items when loading images in useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    The Office
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="abc"
      key="abc"
      title="Gif Title"
      ulr="https://google.com"
    />
  </div>
</Fragment>
`;
